### PR TITLE
Use binary cross entropy when only two classes

### PR DESCRIPTION
**Pull request by theawless**
I was trying to use ktrain in my project for binary text classification. I noticed (on a private dataset) that the model was not converging. I am mostly following the [Text Classification with Hugging Face Transformers](https://nbviewer.jupyter.org/github/amaiya/ktrain/blob/master/tutorials/tutorial-A3-hugging_face_transformers.ipynb) tutorial. As I went through the code, I figured out that CategoricalCrossentropy was the only thing that looked a little out of place, so I tested with BinaryCrossentropy and now my model works much better.

I understand that Binary is just a case of Categorical and it should run fine but it doesn't :(

To test this out further, I ran it on the 20newsgroups data set but I couldn't notice any major difference between the losses.
See [code](https://github.com/theawless/ktrain/commit/db7ef44f72c6c889a57937dda65d100c620d1333) where I try to classify only `comp.graphics` and `soc.religion.christian` classes.

Using BinaryCrossentropy:
```
preprocessing train...
language: en
train sequence lengths:
	mean : 300
	95percentile : 884
	99percentile : 1531
Is Multi-Label? False
preprocessing test...
language: en
test sequence lengths:
	mean : 366
	95percentile : 1046
	99percentile : 3367
Using loss function:  <class 'tensorflow.python.keras.losses.BinaryCrossentropy'>
begin training using onecycle policy with max lr of 8e-05...
Epoch 1/2
99/99 [==============================] - 146s 1s/step - loss: 0.5194 - accuracy: 0.7586 - val_loss: 0.0962 - val_accuracy: 0.9720
Epoch 2/2
99/99 [==============================] - 130s 1s/step - loss: 0.0830 - accuracy: 0.9779 - val_loss: 0.0746 - val_accuracy: 0.9809
```

Using CategoricalCrossentropy:
```
preprocessing train...
language: en
train sequence lengths:
	mean : 300
	95percentile : 884
	99percentile : 1531
Is Multi-Label? False
preprocessing test...
language: en
test sequence lengths:
	mean : 366
	95percentile : 1046
	99percentile : 3367
Using loss function:  <class 'tensorflow.python.keras.losses.CategoricalCrossentropy'>
begin training using onecycle policy with max lr of 8e-05...
Epoch 1/2
148/148 [==============================] - 164s 1s/step - loss: 0.4146 - accuracy: 0.8130 - val_loss: 0.1166 - val_accuracy: 0.9695
Epoch 2/2
148/148 [==============================] - 150s 1s/step - loss: 0.0664 - accuracy: 0.9713 - val_loss: 0.0471 - val_accuracy: 0.9848
```

### CODE REVIEW

**Comment by amaiya** [comment]
Hi @theawless:  Thanks for letting me know about this.  It's strange that you're getting different results on your private dataset.  

However, note that the datasets are preprocessed by default to be multi- or 1-hot-encoded.    If you use BinaryCrossEntropy with  1-hot-encoded labels, you're essentially treating your problem as a multilabel problem (even when it's not supposed to be).  Since this can be [confusing](https://stackoverflow.com/questions/44138324/binary-crossentropy-to-penalize-all-components-of-one-hot-vector) and/or problematic (probabilities will not add to 1), I don't think we should configure models with BinaryCrossEntropy by default for two classes when building model unless perhaps  `preprocess_train`  was also changed to refrain from 1-hot-encoding labels for two classes. However, since, in theory, there  shouldn't be any real different in doing this (as you said) and you are only seeing this behavior in your private dataset, the change probably isn't warranted at this time. (Are you sure the classes in your private dataset are mutually exclusive?)

If you really want to force the model to use BinaryCrossEntropy for your private dataset, you can always just recompile your model with BinaryCrossEntropy:

```python
model = t.get_classifier()
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])
```

... which is essentially equivalent to doing:
```python
model = t.get_classifier(multilabel=True)
```

I'll continue to monitor this.  If it's reproduced in other datasets, it may be a bug in TensorFlow.

Thanks again.

**Comment by theawless** [comment]
Thanks a lot for the quick reply and the informative links.

Turns out my class weights were not getting set up properly (and the dataset was very imbalanced), and that's why the model was not converging for CategoricalCrossentropy. Somehow BinaryCrossentropy worked well even after this.

I fixed the class weights and now CategoricalCrossentropy is also working fine. Thank you for the awesome library!
